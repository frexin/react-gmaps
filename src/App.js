import React from 'react';
import API from './Api'

import MainMap from './components/MainMap'
import MarkersManager from './components/MarkersManager'
import MarkersList from './components/MarkersList'

class App extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            mapCenter: {lat: -3.745, lng: -38.523},
            markers: [],
            isFormActive: false,
            formData: null
        };

        API.get('/markers').then(res => {
            const markers = res.data.data;

            this.setState((res) => ({
                markers: markers
            }));
        });

        this.lastMarkerIndex = null;

        this.handleMarkerSave   = this.handleMarkerSave.bind(this);
        this.handleMarkerDelete = this.handleMarkerDelete.bind(this);
        this.handleMarkerOpen   = this.handleMarkerOpen.bind(this);
        this.handleFormClear    = this.handleFormClear.bind(this);
        this.handleMarkerClick  = this.handleMarkerClick.bind(this);
    }

    handleMarkerSave(marker) {
        if (marker.uid) {
            this.markerUpdate(this.lastMarkerIndex, marker);
        }
        else {
            this.markerCreate(marker);
        }
    }

    handleFormClear() {
        this.setState(() => ({
            formData: null
        }));
    }

    handleMarkerDelete(markerIndex) {
        this.setState((state, props) => ({
            markers: this.state.markers.filter((item, index) => index !== markerIndex)
        }));
    }

    handleMarkerClick(index) {
        const selectedMarker = this.state.markers[index];

        this.setState((state) => ({
            mapCenter: {lat: selectedMarker.lat, lng: selectedMarker.lng}
        }));
    }

    markerUpdate(markerIndex, newData) {
        let updatedMarkers = this.state.markers.map((item, i) => {
            if (i !== markerIndex) {
                return item;
            }

            return newData;
        });

        this.setState((state, props) => ({
            markers: updatedMarkers,
            formData: null,
        }));
    }

    markerCreate(marker) {
        marker.uid = this.state.markers.length + 1;

        this.setState((state, props) => ({
            markers: [...state.markers, marker],
            isFormActive: false,
            formData: null
        }));
    }

    handleMarkerOpen(markerIndex) {
        this.lastMarkerIndex = markerIndex;

        this.setState((state, props) => ({
            formData: state.markers[markerIndex],
        }));
    }

    render() {

        return (
            <div className="container">
                <div className="row">
                    <div className="col">
                        <MainMap mapCenter={this.state.mapCenter} markers={this.state.markers}/>
                    </div>
                    <div className="col">
                        <MarkersManager formActive={this.state.isFormActive} formData={this.state.formData}
                                        saveCallback={this.handleMarkerSave} clearCallback={this.handleFormClear}  />
                        <MarkersList openCallback={this.handleMarkerOpen} deleteCallback={this.handleMarkerDelete}
                                     markers={this.state.markers} clickCallback={this.handleMarkerClick} />
                    </div>
                </div>
            </div>
        );

    }

}

export default App;
