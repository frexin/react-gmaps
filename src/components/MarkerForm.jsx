import React from 'react';
import Geocode from 'react-geocode';

class MarkerForm extends React.Component {

    constructor(props) {
        super(props);

        this.state = this.getDefaultData();

        Geocode.setApiKey("AIzaSyAL3U6qoZFG-zRPHk2kd2Icr7U0lDeVgFE");

        this.handleFormSubmit = this.handleFormSubmit.bind(this);
        this.handleAddressComplete = this.handleAddressComplete.bind(this);
    }

    handleAddressComplete(e) {
        const address = e.target.value;

        if (!address.length) {
            return false;
        }

        Geocode.fromAddress(address).then(
            response => {
                let {lat, lng} = response.results[0].geometry.location;
                this.setState({lat: lat, lng: lng});
            }
        );
    }

    getDefaultData() {
        return {
            lat: '', lng: '', name: '', address: '', uid: ''
        };
    }

    componentWillReceiveProps(nextProps, nextContext) {
        if (nextProps.formData === null) {
            this.setState(() => this.getDefaultData());
        }
        else {
            this.setState(() => nextProps.formData);
        }
    }

    componentDidMount() {
        this.setState(() => this.props.formData);
    }

    handleFormSubmit(event) {
        let elements = event.target.elements;

        let lat = parseFloat(elements.lat.value);
        let lng = parseFloat(elements.lng.value);

        this.setState({
            name: elements.name.value, lat: lat, lng: lng, address: elements.address.value, uid: elements.uid.value
        }, () => {
            this.props.saveCallback(this.state);
        });

        event.preventDefault();
        event.target.reset();
    }

    getClassName() {
        let hide = this.props.isOpen ? '' : ' d-none';
        let name = "marker-form" + hide;

        return name;
    }

    render() {

        return (
            <div className={this.getClassName()}>
                <form onSubmit={this.handleFormSubmit}>
                    <div className="form-group">
                        <label htmlFor="title">Title</label>
                        <input type="text" className="form-control" defaultValue={this.state.name} name="name" id="title"
                        />
                    </div>
                    <div className="form-group">
                        <label htmlFor="addressName">Address</label>
                        <input type="text" onBlur={this.handleAddressComplete} className="form-control" name="address"
                               id="addressName"
                               defaultValue={this.state.address} />
                    </div>
                    <div className="form-row">
                        <div className="form-group col-md-6">
                            <label htmlFor="latitude">Latitude</label>
                            <input type="text" className="form-control" name="lat" id="latitude" defaultValue={this.state.lat}
                                    />
                        </div>
                        <div className="form-group col-md-6">
                            <label htmlFor="longitude">Longitude</label>
                            <input type="text" className="form-control" name="lng" id="longitude" defaultValue={this.state.lng}
                                   />
                        </div>
                    </div>

                    <input type="hidden" name="uid" defaultValue={this.state.uid}  />
                    <button type="submit" className="btn btn-primary">Save</button>
                </form>
            </div>
        )
    }

    static defaultProps = {
        isOpen: false
    }
}

export default MarkerForm
