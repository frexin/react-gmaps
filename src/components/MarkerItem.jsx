import React from 'react';

class MarkerItem extends React.Component {

    render() {

        return (
            <div className="col">
                <h4 className="text-primary">
                    <a href="# " onClick={() => this.props.clickCallback(this.props.index)}>{this.props.name}</a>
                </h4>
                <p className="addrName">{this.props.address}</p>
                <p><b>Long:</b> <span className="lng">{this.props.lng}</span><br/>
                    <b>Lat:</b> <span className="lat">{this.props.lat}</span></p>

                <div className="btn-group" role="group">
                    <button type="button" onClick={() => this.props.deleteCallback(this.props.id)}
                            className="btn btn-danger">Delete</button>
                    <button type="button" onClick={() => this.props.openCallback(this.props.index)}
                            className="btn btn-primary">Edit</button>
                </div>
            </div>
        )
    }

}

export default MarkerItem
