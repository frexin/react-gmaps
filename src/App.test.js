import React from 'react';
import App from './App';

import {mount} from 'enzyme';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';

import MarkerItem from './components/MarkerItem'
import MarkerForm from './components/MarkerForm'

it('renders without crashing', () => {
  shallow(<App/>);
});

describe('Marker Item', () => {
  const marker = {"name": "Test Item", "address": "test addr", "lat": "55.8304", "lng": "49.066", "id": 1};
  const emptyFunc = jest.fn();

  const markerItem = <MarkerItem {...marker} deleteCallback={emptyFunc} openCallback={emptyFunc} />;
  const markerItemComponent = mount(markerItem);

  const output = shallow(markerItem);

  it('render correctly MarkerItem component', () => {
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('component renders correct title', () => {
    expect(markerItemComponent.find('h4 > a').text()).toEqual(marker.name);
    expect(markerItemComponent.find('.addrName').text()).toEqual(marker.address);


    expect(markerItemComponent.find('.lng').text()).toEqual(marker.lng);
    expect(markerItemComponent.find('.lat').text()).toEqual(marker.lat);
  });

  it('component is able to fire event on item click', () => {
    markerItemComponent.find('.btn-danger').first().simulate('click');
    expect(emptyFunc).toHaveBeenCalledWith(1);

    markerItemComponent.find('.btn-primary').first().simulate('click');
    expect(emptyFunc).toHaveBeenCalled();
  })
});

describe('Marker Form', () => {
  const callback = jest.fn();
  let marker = {"name": "Test Item", "address": "test addr", "lat": "55.8304", "lng": "49.066", "id": "1"};

  const markerForm = <MarkerForm formData={marker} saveCallback={callback} />;
  const markerFormComponent = mount(markerForm);

  const output = shallow(markerForm);

  it('render correctly MarkerItem component', () => {
    expect(shallowToJson(output)).toMatchSnapshot();
  });

  it('component renders correct values', () => {
    expect(markerFormComponent.find('#title').instance().value).toEqual(marker.name);
    expect(markerFormComponent.find('#addressName').instance().value).toEqual(marker.address);
    expect(markerFormComponent.find('#latitude').instance().value).toEqual(marker.lat);
    expect(markerFormComponent.find('#longitude').instance().value).toEqual(marker.lng);
  });

  it('component should submit form', () => {
    marker.lat = parseFloat(marker.lat);
    marker.lng = parseFloat(marker.lng);

    markerFormComponent.find('form').first().simulate('submit');
    expect(callback).toHaveBeenCalledWith(marker);
  })
});
