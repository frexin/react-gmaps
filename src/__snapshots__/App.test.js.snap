// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marker Form render correctly MarkerItem component 1`] = `
<div
  className="marker-form d-none"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        className="form-control"
        defaultValue="Test Item"
        id="title"
        name="name"
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="addressName"
      >
        Address
      </label>
      <input
        className="form-control"
        defaultValue="test addr"
        id="addressName"
        name="address"
        onBlur={[Function]}
        type="text"
      />
    </div>
    <div
      className="form-row"
    >
      <div
        className="form-group col-md-6"
      >
        <label
          htmlFor="latitude"
        >
          Latitude
        </label>
        <input
          className="form-control"
          defaultValue="55.8304"
          id="latitude"
          name="lat"
          type="text"
        />
      </div>
      <div
        className="form-group col-md-6"
      >
        <label
          htmlFor="longitude"
        >
          Longitude
        </label>
        <input
          className="form-control"
          defaultValue="49.066"
          id="longitude"
          name="lng"
          type="text"
        />
      </div>
    </div>
    <input
      defaultValue="1"
      name="id"
      type="hidden"
    />
    <button
      className="btn btn-primary"
      type="submit"
    >
      Save
    </button>
  </form>
</div>
`;

exports[`Marker Item render correctly MarkerItem component 1`] = `
<div
  className="col"
>
  <h4
    className="text-primary"
  >
    <a
      href="# "
      onClick={[Function]}
    >
      Test Item
    </a>
  </h4>
  <p
    className="addrName"
  >
    test addr
  </p>
  <p>
    <b>
      Long:
    </b>
     
    <span
      className="lng"
    >
      49.066
    </span>
    <br />
    <b>
      Lat:
    </b>
     
    <span
      className="lat"
    >
      55.8304
    </span>
  </p>
  <div
    className="btn-group"
    role="group"
  >
    <button
      className="btn btn-danger"
      onClick={[Function]}
      type="button"
    >
      Delete
    </button>
    <button
      className="btn btn-primary"
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </div>
</div>
`;
